pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,752 @@
+# see main coroutine function ``nuts_sampler`` for more details
+
+
+@asyncio.coroutine
+def leapfrog(theta, L, grad_L, r, epsilon, inv_mass_matrix):
+    """
+    performs a leapfrog step using a step_size ``epsilon`` and an inverse mass
+    matrix ``inv_mass_matrix``
+
+    The inverse mass matrix can be a 2 dimensional ndarray, in which case it is
+    interpreted as a dense matrix, or a 1 dimensional ndarray, in which case it
+    is interpreted as a diagonal matrix
+    """
+    r_new = r + 0.5 * epsilon * grad_L
+    if inv_mass_matrix.ndim == 1:
+        theta_new = theta + epsilon * inv_mass_matrix * r_new
+    else:
+        theta_new = theta + epsilon * inv_mass_matrix.dot(r_new)
+    L_new, grad_L_new = (yield theta_new)
+    r_new += 0.5 * epsilon * grad_L_new
+    return L_new, grad_L_new, theta_new, r_new
+
+
+@asyncio.coroutine
+def build_tree(state, v, j, adaptor, hamiltonian0, hamiltonian_threshold):
+    """
+    Implicitly build up a subtree of depth j for the NUTS sampler
+    """
+    if j == 0:
+        # Base case - take one leapfrog in the direction v
+        if v == -1:
+            theta = state.theta_minus
+            r = state.r_minus
+            L = state.L_minus
+            grad_L = state.grad_L_minus
+        else:
+            theta = state.theta_plus
+            r = state.r_plus
+            L = state.L_plus
+            grad_L = state.grad_L_plus
+
+        L_dash, grad_L_dash, theta_dash, r_dash = \
+            yield from leapfrog(theta, L, grad_L, r, v * adaptor.get_epsilon(),
+                                adaptor.get_inv_mass_matrix())
+        hamiltonian_dash = L_dash \
+            - kinetic_energy(r_dash, adaptor.get_inv_mass_matrix())
+
+        comparison = hamiltonian_dash - hamiltonian0
+        n_dash = comparison
+        alpha_dash = min(1.0, np.exp(comparison))
+
+        divergent = -comparison > hamiltonian_threshold
+        s_dash = int(not divergent)
+
+        n_alpha_dash = 1
+        return NutsState(
+            theta_dash, r_dash, L_dash, grad_L_dash, n_dash, s_dash,
+            alpha_dash, n_alpha_dash, divergent,
+            adaptor.get_inv_mass_matrix()
+        )
+
+    else:
+        # Recursion - implicitly build the left and right subtrees
+        state_dash = yield from  \
+            build_tree(state, v, j - 1, adaptor, hamiltonian0,
+                       hamiltonian_threshold)
+
+        if state_dash.s == 1:
+            state_double_dash = yield from \
+                build_tree(state_dash, v, j - 1, adaptor, hamiltonian0,
+                           hamiltonian_threshold)
+            state_dash.update(state_double_dash, direction=v, root=False)
+
+        return state_dash
+
+
+@asyncio.coroutine
+def find_reasonable_epsilon(theta, L, grad_L, inv_mass_matrix):
+    """
+    Pick a reasonable value of epsilon close to when the acceptance
+    probability of the Langevin proposal crosses 0.5.
```

Comment:
Is this based on Algorithm 4 in Hoffman? The way in which "r" is sampled looks different?

Reply:
you mean the scaling by the mass matrix? This is as per section 4.2 of [2]

Reply:
Yep, cool -- have added a comment to this effect now.